### PR TITLE
[FLINK-28547][runtime] Add IT cases for SpeculativeScheduler.

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change

Add IT cases for SpeculativeScheduler.

## Verifying this change
`SpeculativeSchedulerITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    private static class NumberCounterSink extends RichSinkFunction<Long> {
+
+        private final Map<Long, Long> numberCountResult = new HashMap<>();
+
+        @Override
+        public void invoke(Long value, Context context) throws Exception {
+
+            numberCountResult.put(value, numberCountResult.getOrDefault(value, 0L) + 1L);
```
Comment (reswqa) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:237]:
```suggestion
            numberCountResult.merge(value, 1L, Long::sum);
```

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+        private boolean hasWaited = false;
+
+        @Override
+        public Long map(Long value) throws Exception {
+            if (toFailCounter.decrementAndGet() >= 0) {
+                throw new Exception("Forced failure for testing");
+            }
+
+            if (!hasWaited) {
```
Comment (zhuzhurk) [inline flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java:215]: We can simplify it as below, the `hasWaited` can be removed then.
```
           if (isSlowTask()) {
                Thread.sleep(1_000_000);
            }
```

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @wanglijie95 
I have a few comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/SpeculativeScheduler.java [hunk 3]
```diff
@@ -254,19 +258,22 @@ private static boolean isExecutionVertexPossibleToFinish(
...
-                        .map(
-                                nodeId ->
-                                        new BlockedNode(
-                                                nodeId,
-                                                "Node is detected to be slow.",
-                                                blockedEndTimestamp))
-                        .collect(Collectors.toList());
-        blocklistOperations.addNewBlockedNodes(nodesToBlock);
+
+        if (!blockSlowNodeDuration.isZero()) {
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/SpeculativeScheduler.java]: Fixed

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    @Test
+    void testBlockSlowNodeInSpeculativeExecution() throws Exception {
+        final Configuration configuration = new Configuration();
+        configuration.set(JobManagerOptions.BLOCK_SLOW_NODE_DURATION, Duration.ofMinutes(1));
+        JobClient client = executeJobAsync(configuration);
+
+        assertThatThrownBy(
+                        () -> client.getJobExecutionResult().get(10, TimeUnit.SECONDS),
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java]: Fixed

File: flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java (hunk 2 above)
Comment (wanglijie95) [reply flink-tests/src/test/java/org/apache/flink/test/scheduling/SpeculativeSchedulerITCase.java]: Fixed

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**